### ISSUE TITLE
requiredIf is broken when using OR field1|field2=1 in complex condition

### ISSUE THREAD

**Issue by hiboudev** [processwire/processwire-issues#1942]
### Short description of the issue

A field 'cars' has this condition in requiredIf (via admin panel):
`teamEvent|mandatoryCarChoice=1,allowCarChoicePerRound=0,carClasses=''`

teamEvent, mandatoryCarChoice and allowCarChoicePerRound: checkboxes
carClasses: multiple page selector (completion)

The red star indicating that the field is required is working fine, but the validation when page is saved it not working in some cases, I don't see error message when I should.
If I activate teamEvent it's OK, but if I activate mandatoryCarChoice it is not working.

Then, on a field 'carClasses' on the same page I set this condition:
`teamEvent|mandatoryCarChoice=1,allowCarChoicePerRound=0,cars=''`

And now nothing is working at all for the validation when page is saved.

If I remove the OR |, letting only teamEvent or mandatoryCarChoice, it's working, so it seems the OR is well managed to display the red star but not for validating the page.

### Expected behavior

When saving the page, if the requiredIf condition is not validated I should see an error message on the empty field (cars or carClasses).

### Setup/Environment

* ProcessWire version: 3.0.240
* (Optional) PHP version: 8.1

**Comment by ryancramerdesign** [comment]
@hiboudev Are these regular fields on the page, or are they embedded within another type of field?

Are the checkboxes fields multiple checkbox fields, or just single checkbox fields? If multiple, you may want the "count" property instead, i.e. `field.count>0`, `field.count=0`, etc. 

Try turning off server-side Inputfield dependency validation temporarily just to confirm that it's a server-side dependency issue, and not something else:
~~~~~
// in /site/config.php: 
$config->InputfieldWrapper = [ 'useDependencies' => false ];
~~~~~

If everything seems to work after doing that, remove that line from config.php, and now try editing the `/wire/modules/Inputfield/InputfieldForm.module` file and change the debug constant at the top to true:
~~~~~
const debug = true; // set to true to enable debug mode for field dependencies
~~~~~
Then edit your page and attempt to reproduce the condition. Save. It should give you a bunch of messages indicating what it validated. We may be able to identify where it's going wrong from these debug messages.

**Comment by hiboudev** [comment]
@ryancramerdesign

> Are these regular fields on the page, or are they embedded within another type of field?

Yes, regular. I use fieldset and fieldset in tab, but I think it's not a problem?

> Are the checkboxes fields multiple checkbox fields, or just single checkbox fields?

Single.

> Try turning off server-side Inputfield dependency validation temporarily just to confirm that it's a server-side dependency issue.

The 2 fields are now always required, whatever the check-boxes statuses, is it expected?

> We may be able to identify where it's going wrong from these debug messages.

Here are the messages with:

- mandatoryCarChoice=1
- teamEvent=0
- allowCarChoicePerRound=0
- carClasses and cars are empty, but the page didn't show any error at save.

Note that there's also a showIf dependency on mandatoryCarChoice: "teamEvent=0", not sure if it's relevant.

![image](https://github.com/processwire/processwire-issues/assets/15386877/b90a3dd2-585c-49e3-ada5-487dfa200a95)

And here are the messages by checking teamEvent (=1) instead of mandatoryCarChoice (=0), now the page is showing an error at save (missing carClasses, missing cars):

![image](https://github.com/processwire/processwire-issues/assets/15386877/449527e7-ef7d-4106-9fb7-f75e57022da2)

**Comment by hiboudev** [comment]
@ryancramerdesign Do you need more information? If you can't understand the issue I could reproduce it in a separate project.

**Comment by matjazpotocnik** [comment]
@ryancramerdesign 

`processInputRequiredIf()` mishandles OR conditions (`field1|field2=value`), while the sibling method `processInputShowIf()` (which drives the correctly-working red star) handles the identical OR syntax correctly. Comparing them side by side exposes the bug directly.

**`processInputShowIf()` (correct — counts matches, OR-safe):**
```php
$numFieldsMatched = 0;
foreach($fields as $name) {
    if($name == '1') {
        $numFieldsMatched++;
    } else if($this->selectorMatchesInputfield($selector, $name, "...")) {
        $numFieldsMatched++;
    }
}
$processNow = $numFieldsMatched > 0; // matches if ANY field in the OR group matched
```

**`processInputRequiredIf()` (buggy — short-circuits on the first non-match):**
```php
foreach($fields as $name) {
    $matches = $this->selectorMatchesInputfield($selector, $name, 'requiredIf'); 
    if($matches === null) continue;
    if($matches === false) {
        $required = false;   // fails the WHOLE OR group on the first field that doesn't match
        break;
    }
}
```

For `teamEvent|mandatoryCarChoice=1`, `$fields` = `['teamEvent', 'mandatoryCarChoice']`. The loop checks each field in turn and bails out to "not required" the moment *any single* field in the OR list fails to match — even if another field in that same OR list *did* match. Since normally only one of the two checkboxes is active at a time, the inactive one almost always kills the whole condition, regardless of which field actually satisfied it. That's exactly the "OK for teamEvent, broken for mandatoryCarChoice" (order/field-dependent) behavior reported.

**Fix** — mirror `processInputShowIf`'s "count matches, OR is satisfied by any one hit" logic while preserving the existing null-handling (field absent from form):**Fix confirmed as a clean, minimal diff** against current dev (`wire/modules/Inputfield/InputfieldForm/InputfieldForm.module`, `processInputRequiredIf()`):

```diff
@@ -447,18 +447,24 @@
 				foreach($selectors as $selector) {
 		
 					$fields = is_array($selector->field) ? $selector->field : array($selector->field);
+					$numFieldsPresent = 0;
+					$numFieldsMatched = 0;
 		
 					foreach($fields as $name) {
 						
 						$matches = $this->selectorMatchesInputfield($selector, $name, 'requiredIf'); 
 						if($matches === null) continue;
-						if($matches === false) {
-							$required = false;
-							break;
-						}
+						$numFieldsPresent++;
+						if($matches === true) $numFieldsMatched++;
 		
 					} // foreach($fields)
 		
+					// an OR'd group of fields (field1|field2=value) is satisfied if ANY one of
+					// them matches; only fail the condition when all present fields failed to match
+					if($numFieldsPresent > 0 && $numFieldsMatched === 0) {
+						$required = false;
+					}
+		
 					if(!$required) break;
 		
 				} // foreach($selectors)
```

**Root cause:** `processInputRequiredIf()` broke the loop and set `$required = false` the moment *any one* field in an OR'd group (`field1|field2=value`) failed to match — instead of only failing when *all* of them fail. Its sibling method, `processInputShowIf()` (which drives the correctly-working red-star display), already does this right by counting matches (`$numFieldsMatched > 0`). The fix just brings `processInputRequiredIf()` in line with that same, already-correct pattern — no new logic invented, just reusing the working approach from the other method.
